### PR TITLE
Add golang messages

### LINKED ISSUE #5: messages
- [x] IngestHeaderChain (addHeaders)
- [x] IngestDifficultyChange (addHeadersWithRetarget)
- [x] MarkNewHeaviest (markNewHeaviest)

### PR DESCRIPTION
**Pull request by ErinHales**
Closes #5

### CODE REVIEW

File: golang/x/relay/keeper/handler.go [hunk 1]
```diff
@@ -4,17 +4,46 @@ import (
...
-		// case types.MsgSetLink:
-		// 	return handleMsgSetLink(ctx, keeper, msg)
+		case types.MsgIngestHeaderChain:
+			return handleMsgIngestHeaderChain(ctx, keeper, msg)
+		case types.MsgIngestDifficultyChange:
+			return handleMsgIngestDifficultyChange(ctx, keeper, msg)
+		case types.MsgMarkNewHeaviest:
+			return handleMsgMarkNewHeaviest(ctx, keeper, msg)
 		default:
 			errMsg := fmt.Sprintf("Unrecognized nameservice Msg type: %v", msg.Type())
```
Comment (prestwich) [inline golang/x/relay/keeper/handler.go]: > nameservice

should say relay. leftover from copy-pasting from elsewhere :rofl:

File: golang/x/relay/keeper/handler.go [hunk 2]
```diff
@@ -4,17 +4,46 @@ import (
...
 		}
 	}
 }
+
+func handleMsgIngestHeaderChain(ctx sdk.Context, keeper Keeper, msg types.MsgIngestHeaderChain) sdk.Result {
+	err := keeper.IngestHeaderChain(ctx, msg.Headers)
+	if err != nil {
+		return err.Result()
+	}
+	return sdk.Result{}
```
Comment (prestwich) [inline golang/x/relay/keeper/handler.go]: modify each of these to say 
```
	return sdk.Result{
		Events: ctx.EventManager().Events(),
	}
```

File: golang/x/relay/types/msgs.go [hunk 3]
```diff
@@ -51,3 +53,133 @@ func (msg MsgSetLink) GetSignBytes() []byte {
...
+		if valid == false || err != nil {
+			return ErrBadHeaderLength(DefaultCodespace)
+		}
+	}
+	return nil
+}
+
+func (msg MsgIngestHeaderChain) GetSignBytes() []byte {
+	return sdk.MustSortJSON(ModuleCdc.MustMarshalJSON(msg))
+
```
Comment (prestwich) [inline golang/x/relay/types/msgs.go]: extra line?

File: golang/x/relay/types/msgs.go [hunk 4]
```diff
@@ -51,3 +53,133 @@ func (msg MsgSetLink) GetSignBytes() []byte {
 
 // Route should return the name of the module
 func (msg MsgSetLink) Route() string { return RouterKey }
+
+/***** IngestHeaderChain *****/
+
+// MsgIngestHeaderChain defines a IngestHeaderChain message
+type MsgIngestHeaderChain struct {
+	Signer  string          `json:"signer"`
```
Comment (prestwich) [inline golang/x/relay/types/msgs.go]: Change the `Signer` type to `sdk.AccAddress` in each message

then implement
```
// GetSigners implements Msg
func (msg MsgWHATEVER) GetSigners() []sdk.AccAddress {
	return []sdk.AccAddress{msg.Signer}
}```

Comment (prestwich) [reply golang/x/relay/types/msgs.go]: that's a bit simpler

Comment (prestwich) [reply golang/x/relay/types/msgs.go]: will have to change the argument type in `func NewWHATEVER` as well

File: golang/x/relay/types/msgs.go [hunk 5]
```diff
@@ -51,3 +53,133 @@ func (msg MsgSetLink) GetSignBytes() []byte {
...
+	return []sdk.AccAddress{addr}
+}
+
+func (msg MsgIngestHeaderChain) Type() string { return "ingest_header_chain" }
+
+func (msg MsgIngestHeaderChain) ValidateBasic() sdk.Error {
+	for i := range msg.Headers {
+		valid, err := msg.Headers[i].Validate()
+		if valid == false || err != nil {
+			return ErrBadHeaderLength(DefaultCodespace)
```
Comment (prestwich) [inline golang/x/relay/types/msgs.go]: I made a `FromBTCSPVError` exactly for this usecase

Comment (prestwich) [reply golang/x/relay/types/msgs.go]: `return FromBTCSPVError(DefaultCodespace, err)`

File: golang/x/relay/types/msgs.go [hunk 6]
```diff
@@ -51,3 +53,133 @@ func (msg MsgSetLink) GetSignBytes() []byte {
...
+	return []sdk.AccAddress{addr}
+}
+
+func (msg MsgIngestDifficultyChange) Type() string { return "ingest_difficulty_change" }
+
+func (msg MsgIngestDifficultyChange) ValidateBasic() sdk.Error {
+	for i := range msg.Headers {
+		valid, err := msg.Headers[i].Validate()
+		if valid == false || err != nil {
+			return ErrBadHeaderLength(DefaultCodespace)
```
Comment (prestwich) [inline golang/x/relay/types/msgs.go]: `return FromBTCSPVError(DefaultCodespace, err)`

File: golang/x/relay/keeper/handler.go (hunk 1 above)
Comment (ErinHales) [reply golang/x/relay/keeper/handler.go]: Good catch, I fixed this in a couple other places as well :D

File: golang/x/relay/types/msgs.go (hunk 3 above)
Comment (ErinHales) [reply golang/x/relay/types/msgs.go]: Oops!

File: golang/x/relay/types/msgs.go (hunk 4 above)
Comment (ErinHales) [reply golang/x/relay/types/msgs.go]: Much better

File: golang/x/relay/types/msgs.go (hunk 5 above)
Comment (ErinHales) [reply golang/x/relay/types/msgs.go]: Ha, right, I found this and then forgot to use it...

File: golang/app.go [hunk 7]
```diff
@@ -190,7 +190,7 @@ func NewRelayApp(logger log.Logger, db dbm.DB, baseAppOptions ...func(*bam.BaseA
 			app.slashingKeeper.Hooks()),
 	)
 
-	// The NameserviceKeeper is the Keeper from the module for this tutorial
+	// The RelayKeeper is the Keeper from the module for this tutorial
 	// It handles interactions with the namestore
```
Comment (prestwich) [inline golang/app.go]: hah, this hsouldn't say namestore :roll_eyes:

File: golang/x/relay/client/cli/query.go [hunk 8]
```diff
@@ -7,9 +7,18 @@ import (
 	"github.com/cosmos/cosmos-sdk/client/context"
 	"github.com/cosmos/cosmos-sdk/codec"
 	"github.com/spf13/cobra"
+	"github.com/summa-tx/bitcoin-spv/golang/btcspv"
 	"github.com/summa-tx/relays/golang/x/relay/types"
 )
 
+type QueryResGetParent struct {
+	Digest btcspv.Hash256Digest `json:"digest"`
```
Comment (prestwich) [inline golang/x/relay/client/cli/query.go]: access as `types.Hash256Digest`

Comment (prestwich) [reply golang/x/relay/client/cli/query.go]: may even be able to omit `types` here? :thinking:

Comment (ErinHales) [reply golang/x/relay/client/cli/query.go]: I missed this earlier, I'll include this in my next PR.  Looks like it has to be `types.`

**Pull request merged**